codec-http2/src/test/java/io/netty/handler/codec/http2/internal/hpack/DecoderTest.java

```diff
@@ -419,4 +423,35 @@ public void testLiteralNeverIndexedWithLargeValue() throws Http2Exception {
         }
         decode(sb.toString());
     }
+
+    @Test
+    public void testDecodeLargerThanMaxHeaderListSizeButSmallerThanMaxHeaderListSizeUpdatesDynamicTable()
+        throws Http2Exception {
+        ByteBuf in = Unpooled.buffer(200);
+        decoder.setMaxHeaderListSize(100, 200);
+        Encoder encoder = new Encoder(true);
+
+        // encode headers that are slightly larger than maxHeaderListSize
+        // but smaller than maxHeaderListSizeGoAway
+        Http2Headers toEncode = new DefaultHttp2Headers();
+        toEncode.add("test_1", "1");
+        toEncode.add("test_2", "2");
+        toEncode.add("long", String.format("%0100d", 0).replace('0', 'A'));
+        toEncode.add("test_3", "3");
+        encoder.encodeHeaders(1, in, toEncode, NEVER_SENSITIVE);
+
+        // decode the headers, we should get an exception, but
+        // the decoded headers object should contain all of the headers
+        Http2Headers decoded = new DefaultHttp2Headers();
+        Http2Exception caught = null;
+        try {
+            decoder.decode(1, in, decoded);
+        } catch (Http2Exception e) {
+            caught = e;
+        }
+        assertTrue(caught instanceof Http2Exception.HeaderListSizeException);
+
```

Comment:
should we also ensure the `in` is handled correctly interns of reference count ?